File: python/0977-squares-of-a-sorted-array.py
# Time: O(n), one pass using two pointers.
# Space: O(1), output array is not considered for space complexity.

class Solution:
    def sortedSquares(self, nums: List[int]) -> List[int]:
        n = len(nums)
        res = [0] * n
        l, r = 0, n - 1
        
        while l <= r:
            left, right = abs(nums[l]), abs(nums[r])
            if left > right:
                res[r - l] = left * left
                l += 1
            else:
                res[r - l] = right * right
                r -= 1
        return res

File: java/0977-squares-of-a-sorted-array.java
class Solution {

    public int[] sortedSquares(int[] nums) {

        int[] result = new int[nums.length];

        int right = nums.length - 1;
        int left = 0;
        int resultIndex = result.length - 1;
        while (left <= right) {

            if (nums[left] * nums[left] >= nums[right] * nums[right]) {
                result[resultIndex] = nums[left] * nums[left];
                left++;
            } else {
                result[resultIndex] = nums[right] * nums[right];
                right--;
            }
            resultIndex--;
        }

        return result;
    }
}